herddb-collections/src/main/java/herddb/collections/TmpMapImpl.java

```diff
@@ -163,39 +187,106 @@ public long estimateCurrentMemoryUsage() {
+    public V get(K key) throws CollectionsException {
         byte[] serializedKey = keySerializer.apply(key);
         return (V) executeGet(serializedKey, tmpTableName);
     }
 
     @Override
-    public boolean containsKey(K key) throws Exception {
+    public boolean containsKey(K key) throws CollectionsException {
         byte[] serializedKey = keySerializer.apply(key);
         return executeContainsKey(serializedKey, tmpTableName);
     }
 
-    private Object executeGet(byte[] serializedKey, String tmpTableName) throws StatementExecutionException, Exception {
-        GetStatement get = new GetStatement(TableSpace.DEFAULT, tmpTableName, Bytes.from_array(serializedKey), null,
-                false);
-        GetResult getResult = (GetResult) tableSpaceManager.executeStatement(get,
-                new StatementEvaluationContext(),
-                herddb.model.TransactionContext.NO_TRANSACTION);
-        if (!getResult.found()) {
-            return null;
-        } else {
-            return valuesSerializer
-                    .deserialize(getResult.getRecord().value);
+    @Override
+    public void forEach(BiSink<K, V> sink) throws CollectionsException, InvocationTargetException {
+
+        try (DataScanner dataScanner =
+                tableSpaceManager.scan(scan, new StatementEvaluationContext(), TransactionContext.NO_TRANSACTION,
+                        false,
+                        false)) {
+            while (dataScanner.hasNext()) {
+                DataAccessorForFullRecord next = (DataAccessorForFullRecord) dataScanner.next();
+                Object key = next.get(0);
+                if (key instanceof RawString) {
+                    key = ((RawString) key).toString();
+                }
+                Object value = valuesSerializer
+                        .deserialize(next.getRecord().value);
+                try {
+                    if (!sink.accept((K) key, (V) value)) {
+                        return;
+                    }
+                } catch (Exception err) {
+                    throw new InvocationTargetException(err);
+                }
+
+            }
+        } catch (InvocationTargetException err) {
+            throw err;
+        } catch (Exception err) {
+            throw new CollectionsException(err);
         }
+
     }
 
-    private boolean executeContainsKey(byte[] serializedKey, String tmpTableName) throws StatementExecutionException,
-            Exception {
-        GetStatement get = new GetStatement(TableSpace.DEFAULT, tmpTableName, Bytes.from_array(serializedKey), null,
-                false);
-        GetResult getResult = (GetResult) tableSpaceManager.executeStatement(get,
-                new StatementEvaluationContext(),
-                herddb.model.TransactionContext.NO_TRANSACTION);
-        return getResult.found();
+    @Override
+    public void forEachKey(Sink<K> sink) throws CollectionsException, InvocationTargetException {
```

Comment:
Why did you used java.lang.reflect.InvocationTargetException and not a CollectionsException specialization like 
"CollectionsInvocationException" or "SinkInvocationException"?